### DISCUSSION TITLE
install multiple versions of pandoc/ reproducible conversions

### DISCUSSION THREAD

**Discussion by denismaier** [jgm/pandoc#10389] [Q&A]
Say I'm using pandoc in conversion pipelines. Is it possible to install multiple versions and switch different ones in different pipelines? Something like python virtualenv or ocaml switches..   that would be nice, because if a certain project depends on pandoc output from a certain version, you might not want to update, but on new projects you will want to use the newest version.

**Comment by tarleb** [comment]
I don't know of any such tool. Personally I use these techniques:

- I wrap all commands in a `Makefile`, define `PANDOC_BIN = pandoc` on top, an then use `$(PANDOC_BIN) ...` in the rules. If it becomes necessary to use a fixed version, then I can change the variable to point to a local pandoc binary, and everything will work again.

- Most of the time, I define a new Dockerfile with a fixed pandoc version (e.g. ,`FROM pandoc/latex:3.5.0-ubuntu`) and use that Docker image to keep track and install all dependencies. It's a bit more work, but makes it very easy to share the build environment, and conversions are sufficiently reproducible.

Truly reproducible builds would probably require Nix or Guix, but I use neither.

**Reply by denismaier** [reply]
> * I wrap all commands in a `Makefile`, define `PANDOC_BIN = pandoc` on top, an then use `$(PANDOC_BIN) ...` in the rules. If it becomes necessary to use a fixed version, then I can change the variable to point to a local pandoc binary, and everything will work again.

Ok, but how do you actually install multiple versions? Just drop the binary somewhere?

**Reply by tarleb** [reply]
I've started [*polytsya*](https://github.com/tarleb/polytsya) and [*verstat*](https://github.com/tarleb/verstat) to make the Docker workflow a little easier when it comes to installing filters. But, like everything recently, it's work in progress and I didn't find the time yet to make it into something that I can announce and recommend in good faith.

**Reply by tarleb** [reply]
> Ok, but how do you actually install multiple versions? Just drop the binary somewhere?

Yes, exactly.

**Reply by tarleb** [reply]
For example, this is what I get when running `ls -l .local/bin/pandoc*`, showing all the pandoc versions that I keep around for testing and other purposes.

```
[I] % ls -l .local/bin/pandoc*
lrwxrwxrwx 1 albert albert       128 Oct  1 17:08 .local/bin/pandoc -> ../../.cabal/store/ghc-9.4.8/pandoc-cli-3.4-e-pandoc-e45d3a374f9e1014ed4b1da2103a541709bc5512d5c304b8826abcf84b25e303/bin/pandoc*
-rwxr-xr-x 1 albert albert 186369360 May  4  2024 .local/bin/pandoc-2.11.0.2*
-rwxr-xr-x 1 albert albert 168997368 May  4  2024 .local/bin/pandoc-2.13*
-rwxr-xr-x 1 albert albert 181652936 May  4  2024 .local/bin/pandoc-2.16.1*
-rwxr-xr-x 1 albert albert 187730152 May  4  2024 .local/bin/pandoc-2.16.2*
-rwxr-xr-x 1 albert albert 188872968 May  4  2024 .local/bin/pandoc-2.17*
-rwxr-xr-x 1 albert albert 189118888 May  4  2024 .local/bin/pandoc-2.17.0.1*
-rwxr-xr-x 1 albert albert 188927656 May  4  2024 .local/bin/pandoc-2.17.1*
-rwxr-xr-x 1 albert albert 235807080 May  4  2024 .local/bin/pandoc-3.1.13*
-rwxr-xr-x 1 albert albert 313422272 Sep 14 09:57 .local/bin/pandoc-3.2-profiler*
-rwxr-xr-x 1 albert albert 316812752 Sep 15 19:33 .local/bin/pandoc-3.4*
-rwxr-xr-x 1 albert albert 128687728 May  4  2024 .local/bin/pandoc-citeproc-0.17.0.2*
-rwxr-xr-x 1 albert albert 113798944 Aug 23 14:34 .local/bin/pandoc-crossref*
-rwxr-xr-x 1 albert albert 115781168 May  4  2024 .local/bin/pandoc-crossref-0.3.8.1*
-rwxrwxr-x 1 albert albert       228 Dec 28  2021 .local/bin/pandocker*
lrwxrwxrwx 1 albert albert         6 Sep 22  2022 .local/bin/pandoc-lua -> pandoc*
-rwxr-xr-x 1 albert albert 188067656 May  4  2024 .local/bin/pandoc-main*
```

**Reply by tarleb** [reply]
Oh, and `pandocker` is a shell script that wraps calls to docker and allows quick access to different versions:

``` sh
#!/bin/sh
version=latest
if [ -n "$PANDOC_VERSION" ]; then
    version="${PANDOC_VERSION}"
fi
 
docker \
    run \
    --rm \
    --volume "$(pwd)":/data \
    --user $(id -u):$(id -g) \
    pandoc/latex:"${version}" \
    "$@"
```

So if I need to use a specific version, I can simply run `PANDOC_VERSION=3.1 pandocker ...`, which is quite convenient.

**Comment by denismaier** [comment]
Cool. Thank you for these answers/suggestions.
